sympy/matrices/expressions/matmul.py

```diff
@@ -333,12 +333,17 @@ def combine_powers(mul):
         if A_base == B_base:
             new_exp = A_exp + B_exp
             new_args[-1] = MatPow(A_base, new_exp).doit(deep=False)
-        elif not isinstance(B_base, MatrixBase) and \
-            A_base == B_base.inverse():
-            new_exp = A_exp - B_exp
-            new_args[-1] = MatPow(A_base, new_exp).doit(deep=False)
-        else:
-            new_args.append(B)
+            continue
+        elif not isinstance(B_base, MatrixBase):
+            try:
```

Comment:
I’ve also discovered that calling inverse for some noninvertible block matrix while canonicalizing had caused some issues.

But can you use ask(Q.invertible) instead of try-catch?

Reply:
Makes sense, thanks. Will do, but it's not trivial, as it breaks at least one test.

Reply:
This ends up being harder than I thought. Doing this breaks tests of the form `A * B == Identity` where `A` and `B` are inverse of each other and where `ask(Q.invertible(B))` just can't provide an answer. Examples are `BlockMatrix([Z]) * BlockMatrix([Z.I])` or `DFT(n) * IDFT(n)` or `KroneckerProduct(W, Z) * KroneckerProduct(W.I, Z.I)`.

I already fixed it for `BlockMatrix` by adding assumption handlers for invertibility. I guess the same can be done for the others. If that all sounds good I'll proceed with this.

(Small note aside: I'm not a fan of invertibility code being in a completely different module than the class being tested. Is there any other way to make `ask(Q.invertible(...))` work? Something like an overridable `MatrixExpr._eval_invertible(self, assumptions)` would sound great.)

Reply:
Typically when the assumptions system returns None, we let things pass through, since it is impossible to make it give True or False in every case.

Although in the cases where we actually need to use the inverse and not just know if it exists, a try/except with `inverse()` is better since it avoids duplicate computation. 

> (Small note aside: I'm not a fan of invertibility code being in a completely different module than the class being tested. Is there any other way to make ask(Q.invertible(...)) work? Something like an overridable MatrixExpr._eval_invertible(self, assumptions) would sound great.)

This gets into bigger design issues with the new assumptions. I would move this to a new issue and not try deal with it in this PR.

Reply:
In the end I agree that try..except is probably safer here. So this code can stay as is I guess.

Reply:
Is the except only catching exceptions from the call to inverse? If so then the other code should not be in the try block.

Reply:
Good point.